### ISSUE TITLE
Nice stuffs

### ISSUE THREAD

**Issue by finardi** [mapmeld/ml#1]
Hey @mapmeld like a lot of your texts about NLP and ML in general, also I read some mediums the you wrote, very nice. Did you improve your results in ELECTRA? I'm from Brazil, I got a nice dataset 18GB, one line per sample, right now I'm building the build_pretraining_dataset.py (taking 4hours... until now). 

I'll use GCS with colab, hope this work, will be my first time. 
If it's possible clarify some doubts that I have about TPU training:

1. Do I need to move my 18gb dataset to GCS? By now I the data is loadad in google drive. 
2. I'm building the pretrain_tfrecords on google drive, same folder of my dataset, do I need move it to GCS?
3. How to load a checkpoint in GCS? I did some tests with small data only in colab (gpu) and this is pretty automatic, I mean I run for 200 steps, reset everything and start again (at least the loss start in almost same point). 

Thank you for your time, wish the best!

**Comment by mapmeld** [comment]
Hi!  The latest version of my code (including successful training on TPU) is here: https://colab.research.google.com/drive/1R8TciRSM7BONJRBc9CBZbzOmz39FTLl_?usp=sharing
If you are more familiar with PyTorch or are frustrated by the ELECTRA docs, there is a new re-implementation of ELECTRA, you can try: https://twitter.com/_RichardWang_/status/1302545459374772224

I have been using CoLab Pro for GPU+TPU access. I recommend CoLab because you can try GPU training there for free until it is working, and then move to other hardware when you feel confident

The TPU run took a few changes which you see in the notebook already; 
- build the pretrain_tfrecords with the argument `--max-seq-length 512` - and `--no-lower-case` if you have case-sensitive vocab.txt
- copy pretrain_tfrecords folder to your Google Drive. If you download the directory from Google Drive, make sure you get ALL 1,000 tfrecord files.  Google Drive divided them into two ZIP downloads so I messed up at first
- Create a Google Storage bucket with permissions open to your machine or CoLab. The top-level bucket has: vocab.txt, the pretrain_tfrecords directory, and a directory named "models". Your model will be stored in the "models/model_name" directory
- For CoLab, run `print(os.environ['TPU_NAME'])` to find your assigned TPU
- For run_pretraining.py use arguments with your values replacing CAPITALIZED terms here `--data-dir gs://BUCKET_NAME --model-name MODEL_NAME --hparams '{"model_size": "base", "use_tpu": true, "num_tpu_cores": 8, "max_seq_length": 512, "tpu_name": "grpc://COLAB_TPU_ADDRESS", "vocab_size":VOCAB_FILE_SIZE}'`
- Your model will make a lot of noise if something is missing. It should save a checkpoint every 1,000 steps.  I am not sure why yours did not resume from the model_name directory as planned

**Comment by finardi** [comment]
Hey @mapmeld thanks a lot for the insights!!! 

I follow your steps, I believe it's ok by now, here my notebook: https://colab.research.google.com/drive/1nwmcbEm7uq_rVxFTp3QBDQ9kobkeFT00?usp=sharing (the build_pretraining_dataset and the tokenizer, was not present in this colab, but It's pretty much like yours). Few Questions:

**1.** About the output, Do you know what's _SPS_, for example (after 2.9%) "29000/1000000 = 2.9%, _SPS_: 5.1, ELAP: 1:24:42, ETA: 2 days, 4:43:10"

**2.** In your code you saved the tokenizer in json format (but if you just save it's a .txt), why do you do this:

```
import json
layout = json.loads(open('./vocab.json', 'r').read())
op = open('./vocab.txt', 'w')
for word in layout["model"]["vocab"].keys():
  op.write(word + "\n")
```
Is it kind of some trick?

**3.** When i trained with GPU the output was showing the loss, with TPU I cant find the loss. I'm missing some point in my colab?

Thanks for advance, pretty cool talk to you about this, cheers!

**Comment by mapmeld** [comment]
1. SPS = steps per second ( https://github.com/google-research/electra/blob/81f7e5fc98b0ad8bfd20b641aa8bc9e6ac00c8eb/util/training_utils.py#L68 ) - in my experience it starts at different speeds and then converges at one constant speed
2. I might have read some weird docs. Yes better to save just as a vocab.txt
3. ELECTRA skips calculating the loss during TPU training. I tried editing this, but it got complicated. I don't know why they did this https://github.com/google-research/electra/blob/79111328070e491b287c307906701ebc61091eb2/run_pretraining.py#L288

**Comment by finardi** [comment]
Thanks for the reply! Are you using transformers HuggingFaces to convert tensorflow to PyTorch? I found this colab 
https://colab.research.google.com/drive/1DiOwhRjQbtYRgFWG7e3dybcXJsZcu86l

After the pre-training the guy convert tf to pytorch with this repo: https://github.com/lonePatient/electra_pytorch.git
Idk which is the best strategy here. 


Also, after the pre-training I'm thinking to setup a fine-tuning classification task in Pytorch, something like this

```
class ElectraClass(torch.nn.Module):
    def __init__(self, num_labels=2):
        super(ElectraClass, self).__init__()

        self.num_labels=num_labels
        self.l1 = ElectraModel.from_pretrained(MODEL-PRETRAINED, return_dict=True)
        self.pre_classifier = torch.nn.Linear(768, 768)
        self.dropout = torch.nn.Dropout(0.1)
        self.classifier = torch.nn.Linear(768, num_labels)

    def forward(self, input_ids, token_type_ids, attention_mask, labels=None):
        outs = self.l1(input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask)
        hidden_state = outs['last_hidden_state']
        pooler = hidden_state[:,0]
        pooler = self.pre_classifier(pooler)
        pooler = torch.nn.Tanh()(pooler)
        
        pooler = self.dropout(pooler)
        output = self.classifier(pooler)
        
        if labels is not None:
            loss_fct = torch.nn.CrossEntropyLoss()
            loss = loss_fct(output.view(-1, self.num_labels), labels.view(-1))
            output = {'loss':loss, 'logits':output}
            return output 
        
        return output
        
```

Any ideas to convert tf to pytorch and perform fine-tunning? This is my first pretraining from scratch, learning a lot :)

**Comment by mapmeld** [comment]
I see they are using `electra_pytorch`, that is new to me, looks like it will be a good method. My current (messy, maybe outdated) way of getting a pytorch_model.bin and a tf_model.h5

```python
from transformers import BertTokenizer
tkn = BertTokenizer('./hindi-bert/vocab.txt', strip_accents=False, lowercase=False)
tkn.save_pretrained('./hindi-bert/') # outputs vocab.json
```

Make a valid config.json based on https://github.com/huggingface/transformers/issues/3671 and other HF repos

Converting checkpoint to PyTorch:

```
python3 transformers/src/transformers/convert_electra_original_tf_checkpoint_to_pytorch.py
    --tf_checkpoint_path=./models/hindi-bert
    --config_file=config.json
    --pytorch_dump_path=pytorch_model.bin
    --discriminator_or_generator=discriminator
```

Converting PyTorch back to h5 for TensorFlow

```python
from transformers import TFElectraForPreTraining
model = TFElectraForPreTraining.from_pretrained("./hindi-bert", from_pt=True)
model.save_pretrained("tf-output")
```

**Comment by finardi** [comment]
Hey @mapmeld I tested the model with 45% of training, approximately  330K steps. 
I'm using your tip to convert TF to PyTorch (the HuggingFaces script: `convert_electra_original_tf_checkpoint_to_pytorch.py`) and the `config.json` I build like bellow, following the table6 of the Electra-Paper, in this point idk if I messed up the numbers.  

```
config = {
  "attention_probs_dropout_prob": 0.1,
  "model_type": 'electra',
  "hidden_size": 768,
  "intermediate_size": 3072,
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "embedding_size": 768,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "initializer_range": 0.02,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "type_vocab_size": 2,
  "vocab_size": 34100
}
with open("/content/config.json", "w") as f:
    json.dump(config, f)
```
After this, I use

``` 
tokenizer = AutoTokenizer.from_pretrained('/content')
model = AutoModel.from_pretrained('/content', return_dict=True)
```
With the same class in previous post ( ```ElectraClass```) got low convergence and low score compared with the HuggingFaces Electra (tested the small and base versions). Maybe my low convergence is because 330k steps? I'll train until 766k steps (like the paper).

**Comment by mapmeld** [comment]
This is the config.json which I used recently on the TPU / base size model: https://s3.amazonaws.com/models.huggingface.co/bert/monsoon-nlp/hindi-tpu-electra/config.json - looks similar

I don't know enough about it to say what would improve the model. What was the size of your original corpus?

**Comment by mapmeld** [comment]
I am thinking about using electra_pytorch going forward because it's documented. But you won't be able to use a TPU, with PyTorch, right? https://github.com/richarddwang/electra_pytorch/blob/master/Pretrain.ipynb

**Comment by finardi** [comment]
My corpus has 16gb, I'm using for the tokenizer wikipedia sentences ~ 3mi sentences. About electra_pytorch looks nice, but no TPU I think. 

Maybe I should try to train BERT, it seems easier. What do you think? But, I'm not sure which implementation to use, official/Hugging Faces... Any tips for BERT from scratch?

Thanks again @mapmeld !

**Comment by mapmeld** [comment]
Sorry no tips, this is as far as I've gotten on my own projects.

**Comment by mapmeld** [comment]
An explanation on TPUs, PyTorch, and measuring loss during TPU training: https://twitter.com/KaliTessera/status/1321454546569338882
